File: src/wolfcom.java
package com.zdiv.jlib.app.WebToon;

import java.io.File;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import com.zdiv.jlib.base.Encoding;
import com.zdiv.jlib.base.FileUtility;

public class WolfCom {
	
	final static boolean debug = false;
	final static String filter = null; //"jpg";
	
	public static Document getJsoupDocument(String url) throws InterruptedException {
		while( true ) {
			try {
				return Jsoup.connect(url).get();
			} catch ( Exception e ) {
				e.printStackTrace();
				Thread.sleep(1000);
			}
		}
	}


	public static void downloadFile(String urlStr, String fileName, String referer) throws IOException {
        URL url = new URL(urlStr);
        HttpURLConnection  hc = (HttpURLConnection) url.openConnection();
        hc.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36");
        hc.setRequestProperty("Referer", referer);
       
    	int status = hc.getResponseCode();
    	while (status != HttpURLConnection.HTTP_OK) {
    		if (status == HttpURLConnection.HTTP_MOVED_TEMP
    			|| status == HttpURLConnection.HTTP_MOVED_PERM
    			|| status == HttpURLConnection.HTTP_SEE_OTHER ) {
        		String newUrl = hc.getHeaderField("Location");
        		hc = (HttpURLConnection) new URL(newUrl).openConnection();
        		hc.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36");
            	status = hc.getResponseCode();
    		}
    		if (status == HttpURLConnection.HTTP_NOT_FOUND ) {
    			return;
    		}
    	}            
        
        Files.copy(hc.getInputStream(), Paths.get(fileName), StandardCopyOption.REPLACE_EXISTING);
	}
	

	public static void getWolfCom(String comicsUrl, String baseUrl, String baseDir)  
			throws InterruptedException, MalformedURLException, IOException {

		Document doc_toc = getJsoupDocument(comicsUrl);
		if( debug ) {
			//String text = doc_toc.text();
			String html = doc_toc.html();
			System.out.println(html);
			FileUtility.StringToFile("D:/aa.html",html);
		}
		
		Elements list = doc_toc.select("div.box > div.group.left-box > div.webtoon-bbs-list.bbs-list > ul > li");
		if( debug ) {
			System.out.println(list.html());
		}
		
		File dir = new File(baseDir,doc_toc.title().replaceAll("[?/:]","_"));
		dir.mkdirs();
	
		int i = 0;
		for( Element e : list ) {
			
			//if( i++ < 33 ) continue;
			//if( ++i > 3 ) break;
			
			try {
				String url = baseUrl + e.select("a").first().attr("href");
				if( debug ) {
					System.out.println(url);
				}
				
				Document doc_img = getJsoupDocument(url);
				Elements imgs = doc_img.select("section.webtoon-body > div.group.image-view > img");
				System.out.println(doc_img.title());
				
				File subdir = new File(dir.getPath(),doc_img.title().replaceAll("[?/:]","_"));
				if( subdir.exists() ) {
					continue;
				} else {
					subdir.mkdirs();
				}
				
				int k = 1;
				for( Element img : imgs ) {
					String img_url = img.attr("src");
					if( filter == null || img_url.endsWith(filter) ) { 
						if( ! img_url.startsWith("http") ) {
							img_url = baseUrl + img_url;
						}
						String file_name = String.format("img_%04d.jpg",k++);
						System.out.println( img_url + " -> " + file_name );
						downloadFile(img_url,subdir.getPath() + "/" + file_name,comicsUrl);
					}
				}
			} catch( Exception e1 ){
			
			}
			//break;
		}
	}
	
	public static void main(String[] args) throws InterruptedException, MalformedURLException, IOException {
		String[] url = {
				//"https://wfwf104.com/list?toon=1229",
				"https://wfwf105.com/list?toon=381&title=%B8%BE%C4%AB%C6%E4",
			};
		String iurl = "https://wfwf105.com";
		String dir = "D:/Temp4/";
		for( String u : url ) {
			getWolfCom(u, iurl, dir);
			System.out.println( "END" );
		}
	}
}

File: src/wolfcom.py
import bs4, codecs
import requests
import os
import io

requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)
image_ext = 'jpg'
request_headers = {
    'User-Agent': ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 '
                   '(KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'),
    'Referer' : ("https://wfwf105.com/list?toon=381&title=%B8%BE%C4%AB%C6%E4")
}

def getFile(url):
    with codecs.open(url,'r', encoding='utf8') as f:
        html = f.read()
    return bs4.BeautifulSoup(html, 'html.parser')

def getUrl(url, headers={}, params=()):
    resp = requests.get(url, verify=False, headers=headers, params=params)
    #resp.headers
    #html = resp.content.decode('utf8')
    html = resp.text
    return bs4.BeautifulSoup(html, 'html.parser')

def urlToFile(url, file_name):
    resp = requests.get(url, verify=False, headers=request_headers, params=())
    with open(file_name, "wb") as f:
        f.write(resp.content)

def extractTag(bs,tag):
    [s.extract() for s in bs(tag)]

def getWolfCom( comicsUrl, baseUrl, baseDir):
    doc = getUrl(comicsUrl)
    title = doc.find("title").text
    elist = doc.select("div.box > div.group.left-box > div.webtoon-bbs-list.bbs-list > ul > li")
    
    new_dir = os.path.join(baseDir, title.replace(":","_"))
    if not os.path.isdir(new_dir): os.mkdir(new_dir)

    for e in elist:
        a = e.find('a',"view_open",href=True)
        if not a: continue
        url = baseUrl + a['href']

        doc = getUrl(url, request_headers)
        title = doc.find("title").text
        imgs = doc.select("section.webtoon-body div.group.image-view img")

        sub_dir = os.path.join(new_dir, title.replace(":","_"))
        if not os.path.isdir(sub_dir): os.mkdir(sub_dir)

        k = 1;
        for img in imgs:
            img_url = img.get('src')
            if not img_url: continue
            if image_ext == None or img_url.endswith(image_ext):
                if( not img_url.startswith("http") ):
                    img_url = baseUrl + img_url
                file_name = "img_%04d.jpg" % k
                urlToFile( img_url, os.path.join( sub_dir, file_name) )
                print( img_url + " -> " + file_name )
                k = k + 1

if __name__ == "__main__":
        url = "https://wfwf105.com/list?toon=381&title=%B8%BE%C4%AB%C6%E4"
        iurl = "https://wfwf105.com"
        bdir = "D:/Temp2/"
        getWolfCom(url, iurl, bdir)

